### ISSUE TITLE
SQLFluff produces corrupt code with Jinja whitespace trimming

### ISSUE THREAD

**Issue by CyberShadow** [sqlfluff/sqlfluff#1321]
## Input

```sql
-- Beginning of file, two blank lines follow


{%- set x = 42 %}
SELECT 1, 2

```

(the blank lines are important; the comment is just to prevent the GitHub Markdown renderer from trimming the leading blank lines)

## Command

```console
$ sqlfluff fix --force test.sql
```

## Expected output

Probably something like:

```sql
-- Beginning of file, two blank lines follow


{%- set x = 42 %}
SELECT
    1,
    2
```

## Actual output

```sql
-- Beginning of file, two blank lines follow


{%- set x = 42 SELECT
    1,
    2, 2
```

## Version
Reproducible with SQLFluff 0.6.4 and master on Python 3.9.6

## Configuration

None.

**Comment by barrywhart** [comment]
Looking at this in the debugger, I see that no matter how many blank lines appear at the beginning of the file, at this line in `linted_file.py`, `patch.templated_slice` is `slice(1, 12, None)`. This suggests to me that _maybe_ the white space at the beginning of the file is being discarded before the position in the file is computed.

```
                source_slice = self.templated_file.templated_slice_to_source_slice(
                    patch.templated_slice,
```

**Comment by barrywhart** [comment]
Related: See file slice 1 in the detailed log output below. The `source` field has the value `'%}\nSELECT 1, '`.

```
==== fixing violations ====
1 fixable linting violations found
Are you sure you wish to attempt to fix these? [Y/n] ...
Attempting fixes...
Persisting Changes...
DEBUG      Original Tree: '\nSELECT 1, 2\n'
DEBUG      Fixed Tree: '\nSELECT\n    1,\n    2\n'
DEBUG      ### Templated File.
DEBUG          File slice: 0 TemplatedFileSlice(slice_type='templated', source_slice=slice(0, 17, None), templated_slice=slice(0, 0, None))
DEBUG          			templated: ''	source: '\n\n{%- set x = 42 '
DEBUG          File slice: 1 TemplatedFileSlice(slice_type='literal', source_slice=slice(17, 30, None), templated_slice=slice(0, 13, None))
DEBUG          			templated: '\nSELECT 1, 2\n'	source: '%}\nSELECT 1, '
```

**Comment by barrywhart** [comment]
I made some possible progress. Adding the following lines to `TemplatedFile.__init__` fixed the issue listed here. However, it broke a number of other tests, which I haven't looked into yet. 
```
        import re
        m = re.search(r'^\s*', source_str)
        if m:
            source_str = source_str[len(m.group(0)):]
```

It seems that a major contributor to this issue is that the Jinja `Template` class discards leading whitespace at the beginning of a template. I don't see a way to prevent this behavior (although there **is** a way to preserve **trailing** whitespace). Thus, pre- and post-templated files become misaligned, and SQLFluff gets confused when it tries to find the matching areas of the "before" and "after" contents. The above change attempts to address this by discarding leading whitespace early.

I'll work on this issue some more if I have time, but if anyone else is interested, please go ahead. I don't know if/when I'll be able to return to this.

**Comment by barrywhart** [comment]
I think I've got a fix for this. Working through the affected test cases, mostly Jinja templating stuff. I'll assign the issue to myself, on the assumption I'll finish it. 🤞🏽😅

**Comment by barrywhart** [comment]
Confirmed that this issue is not a duplicate of #854. PR #1364 fixes this issue but not that one.

**Comment by barrywhart** [comment]
Correction: The two issues **are** duplicates. Now #1364 fixes them both.

**Comment by CyberShadow** [comment]
Closing as #1364 is now merged. Thank you!
